Mathlib/Geometry/Euclidean/Similarity.lean

```diff
@@ -0,0 +1,278 @@
+  set k := (dist a b / dist a' b') with hk
+  have k_ne : k > 0 := by
+    rw [hk]
+    apply div_pos
+    · simp [dist_pos, ne₁₂_of_not_collinear h_not_col]
+    · simp [dist_pos, ne₁₂_of_not_collinear h_not_col']
+  have h_ab : dist a b = k * dist a' b' := by
+    rw [hk, div_mul, div_self dist_a'b', div_one]
+  have h_bc : dist b c = k * dist b' c' := by
+    rw [hd, div_mul, div_self dist_b'c', div_one]
+  have hcos := law_cos a b c
+  rw [dist_comm b _ , dist_comm b' _] at h_bc
+  rw [h_ab, h_bc] at hcos
+  field_simp at hcos
+  rw [h] at hcos
+  have hcos' := law_cos a' b' c'
+  field_simp at hcos'
+  rw [← hcos', ← mul_pow] at hcos
+  have dist_ac : 0 < dist a c := by
+    simp [dist_pos, ne₁₃_of_not_collinear h_not_col]
+  have dist_a'c' : 0 < dist a' c' := by
+    simp [dist_pos, ne₁₃_of_not_collinear h_not_col']
+  have k_dist_a'c' : 0 ≤ k * dist a' c' := by positivity
+  rw [pow_left_inj₀ (le_of_lt dist_ac) k_dist_a'c' (by norm_num)] at hcos
+  rw [dist_comm c _, dist_comm c' _] at h_bc
+  rw [dist_comm a _, dist_comm a' _] at hcos
+  apply similar_of_side_side_side
+  use k
+
+/-- For two similar triangles, there exists a positive ratio such that the distances between
+corresponding vertices are proportional. -/
+theorem exist_dist_eq_mul_dist_of_similar {t₁ : Fin 3 → P₁} {t₂ : Fin 3 → P₂} (h : t₁ ∼ t₂)
+    (i j : Fin 3) :
+    ∃ r : ℝ, r > 0 ∧ dist (t₁ i) (t₁ j) = r * dist (t₂ i) (t₂ j) := by
+  rw [triangle_similar_iff_exists_dist_eq] at h
+  rcases h with ⟨r, hr_pos, hdist⟩
+  use r
+  exact ⟨hr_pos, hdist i j⟩
+
+/-- For two similar triangles, the corresponding angles are equal. -/
+theorem angle_eq_of_similar (h : ![a, b, c] ∼ ![a', b', c']) :
+    ∠ a b c = ∠ a' b' c' := by
+  rw [triangle_similar_iff_exists_dist_eq] at h
+  rcases h with ⟨r, hr_pos, hdist⟩
+  have h_ab : dist a b = r * dist a' b' := hdist 0 1
+  have h_cb : dist c b = r * dist c' b' := hdist 2 1
+  have h_ac : dist a c = r * dist a' c' := hdist 0 2
+  have h_cos := law_cos a b c
+  rw [h_ab, h_cb, h_ac] at h_cos
+  field_simp at h_cos
+  have h_cos' := law_cos a' b' c'
+  field_simp at h_cos'
+  rw [h_cos', sub_right_inj] at h_cos
+  by_cases heq : dist a' b' * dist c' b' * 2 = 0
+  · rw [mul_eq_zero_iff_right (by norm_num), mul_eq_zero] at heq
+    rcases heq with h1 | h2
+    · have h_dist_ab : dist a b = 0 := by
+        rw [h_ab, h1, mul_zero]
+      rw [dist_eq_zero] at h_dist_ab h1
+      simp_rw [h_dist_ab, h1, angle_self_left]
+    · have h_dist_cb : dist c b = 0 := by
+        rw [h_cb, h2, mul_zero]
+      rw [dist_eq_zero] at h_dist_cb h2
+      simp_rw [h_dist_cb, h2, angle_self_right]
+  rw [mul_right_inj' heq] at h_cos
+  apply Real.injOn_cos at h_cos
+  · rw [h_cos]
+  repeat exact ⟨angle_nonneg _ _ _, angle_le_pi _ _ _ ⟩
+
+/-- Reindexing both triangles by the same permutation preserves similarity. -/
+theorem similar_reindex {t₁ : ι → P₁} {t₂ : ι → P₂} (e : Equiv.Perm (ι)) :
+    t₁ ∼ t₂ ↔ (t₁ ∘ e) ∼ (t₂ ∘ e) := by
```

Comment:
This isn't Euclidean, it belongs in the generic file, for *two* arbitrary index types with an equiv between them rather than requiring a permutation of a single index type. (And it should be in the `PseudoEMetricSpace` part of that file; it shouldn't need anything more than that.)